kernel/aster-nix/src/fs/utils/dentrymnt.rs

```diff
@@ -0,0 +1,191 @@
+                        let slow_dentry = self.dentry.lookup_slow(name)?;
+                        Self::new(self.mount_node().clone(), slow_dentry.clone())
+                    }
+                }
+            }
+        };
+        let dentrymnt = dentrymnt.overlaid_dentrymnt();
+        Ok(dentrymnt)
+    }
+
+    // Get the absolute path.
+    //
+    // It will resolve the mountpoint automatically.
+    pub fn abs_path(&self) -> String {
+        let mut path = self.effective_name();
+        let mut dir_dentrymnt = self.this();
+
+        loop {
+            match dir_dentrymnt.effective_parent() {
+                None => break,
+                Some(parent_dir_dentrymnt) => {
+                    path = {
+                        let parent_name = parent_dir_dentrymnt.effective_name();
+                        if parent_name != "/" {
+                            parent_name + "/" + &path
+                        } else {
+                            parent_name + &path
+                        }
+                    };
+                    dir_dentrymnt = parent_dir_dentrymnt;
+                }
+            }
+        }
+        debug_assert!(path.starts_with('/'));
+        path
+    }
+
+    /// Get the effective name of dentrymnt.
+    ///
+    /// If it is the root of mount, it will go up to the mountpoint to get the name
+    /// of the mountpoint recursively.
+    fn effective_name(&self) -> String {
+        if !self.dentry.is_root_of_mount() {
+            return self.dentry.name();
+        }
+
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
+            let parent_dentrymnt = Self::new(
+                self.mount_node.parent().unwrap().upgrade().unwrap().clone(),
+                self.mount_node.mountpoint_dentry().unwrap().clone(),
+            );
+            parent_dentrymnt.effective_name()
+        } else {
+            self.dentry.name()
+        }
+    }
+
+    /// Get the effective parent of dentrymnt.
+    ///
+    /// If it is the root of mount, it will go up to the mountpoint to get the parent
+    /// of the mountpoint recursively.
+    fn effective_parent(&self) -> Option<Arc<Self>> {
+        if !self.dentry.is_root_of_mount() {
+            return Some(Self::new(
+                self.mount_node.clone(),
+                self.dentry.parent().unwrap().clone(),
+            ));
+        }
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
+            let parent_dentrymnt = Self::new(
+                self.mount_node.parent().unwrap().upgrade().unwrap().clone(),
+                self.mount_node.mountpoint_dentry().unwrap().clone(),
+            );
+            parent_dentrymnt.effective_parent()
+        } else {
+            None
+        }
+    }
+
+    /// Get the overlaid dentrymnt of self.
+    ///
+    /// It will jump into the child mount if it is a mountpoint.
+    fn overlaid_dentrymnt(&self) -> Arc<Self> {
```

Comment:
I find it abrupt to see this word of "overlaid". The verb "overlay" does not appear anywhere in our current codebase or is defined by our Rust doc. The first idea that comes to my mind when I see this word is overlayfs. But apparently, the code has nothing to do with overlayfs. Could you think of a more appropriate word?

Reply:
When multiple file systems are mounted on the same mount point, consider the following scenario: We sequentially execute the following three commands:
```shell
mount /dev/sda1 /mnt
mount /dev/sda2 /mnt
mount /dev/sda3 /mnt
```
Upon mounting `/dev/sda1` to `/mnt`, we can access the contents of `/dev/sda1` through the /mnt directory. When `/dev/sda2` is also mounted to `/mnt`, the contents of `/dev/sda1` are overshadowed (not replaced) by `/dev/sda2`. At this point, the contents accessible at `/mnt` are from `/dev/sda2`. Similarly, when `/dev/sda3` is mounted to `/mnt`, the contents of `/dev/sda2` are overshadowed by `/dev/sda3`, and the contents accessible at `/mnt` are now from `/dev/sda3`.

Unmounting the file system mounted at `/mnt` (which is `/dev/sda3` at the top) allows us to access the contents of `/dev/sda2` at `/mnt`. If we unmount the file system from `/mnt` again, we can then access the contents of `/dev/sda1`.

In this example, the file systems behave as if they are placed in a stack. The file system we can access at `/mnt` at any given time is the one at the top of the stack. The original `overlaid_dentrymnt` was used to retrieve the file system at the top. Since we always start our search from the bottom, to more clearly represent this process, I have renamed `overlaid_dentrymnt` to `get_top_dentrymnt`, which signifies obtaining the `DentryMnt` of the file system at the top of the stack.

Reply:
`get_top_dentrymnt` sounds better 👍